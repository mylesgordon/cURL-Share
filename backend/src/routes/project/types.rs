use actix_session::SessionGetError;
use actix_web::HttpResponse;
use serde::{Deserialize, Serialize};

use crate::routes::types::UserError;

#[derive(Debug)]
pub enum ProjectError {
    Forbidden(String),
    ProjectDoesNotExistError(String),
    SessionGetError(String),
    SqlxError(sqlx::Error),
    UserError(UserError),
}

impl From<SessionGetError> for ProjectError {
    fn from(e: SessionGetError) -> Self {
        match e {
            // TODO: improve
            _ => ProjectError::SessionGetError(e.to_string()),
        }
    }
}

impl From<sqlx::Error> for ProjectError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => {
                ProjectError::ProjectDoesNotExistError("Project was not found".to_string())
            }
            _ => ProjectError::SqlxError(e),
        }
    }
}

impl From<UserError> for ProjectError {
    fn from(e: UserError) -> Self {
        match e {
            // TODO: improve
            UserError::SessionGetError(e) => ProjectError::SessionGetError(e),
            _ => ProjectError::UserError(e),
        }
    }
}

impl From<ProjectError> for HttpResponse {
    fn from(e: ProjectError) -> Self {
        match e {
            ProjectError::ProjectDoesNotExistError(_) => HttpResponse::NotFound().finish(),
            ProjectError::SessionGetError(_) => HttpResponse::Unauthorized().finish(),
            ProjectError::Forbidden(_) => HttpResponse::Forbidden().finish(),
            _ => HttpResponse::InternalServerError().finish(),
        }
    }
}

#[derive(Serialize)]
pub struct ProjectId {
    pub id: i64,
}

#[derive(Debug, Default, Deserialize)]
pub struct ProjectParams {
    pub search: String,
    pub labels: Vec<String>,
}
