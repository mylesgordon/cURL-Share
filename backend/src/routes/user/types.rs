use actix_web::HttpResponse;
use secrecy::Secret;
use serde::Deserialize;

#[derive(Debug)]
pub enum UserError {
    PasswordHashError(argon2::password_hash::Error),
    SqlxDatabaseError,
    SqlxError(sqlx::Error),
    UserAlreadyExists,
    UserNotFound(sqlx::Error),
}

impl From<argon2::password_hash::Error> for UserError {
    fn from(e: argon2::password_hash::Error) -> Self {
        Self::PasswordHashError(e)
    }
}

impl From<sqlx::Error> for UserError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => Self::UserNotFound(e),
            sqlx::Error::Database(database_err) => {
                if let Some(database_err_code) = database_err.code() {
                    match database_err_code.as_ref() {
                        "2067" => Self::UserAlreadyExists,
                        _ => Self::SqlxDatabaseError,
                    }
                } else {
                    Self::SqlxDatabaseError
                }
            }
            _ => Self::SqlxError(e),
        }
    }
}

impl From<UserError> for HttpResponse {
    fn from(e: UserError) -> Self {
        match e {
            UserError::UserAlreadyExists => HttpResponse::Conflict().into(),
            UserError::UserNotFound(_) => HttpResponse::Unauthorized().into(),
            _ => HttpResponse::InternalServerError().into(),
        }
    }
}

#[derive(Deserialize)]
pub struct UserRequest {
    pub username: String,
    pub password: Secret<String>,
}
