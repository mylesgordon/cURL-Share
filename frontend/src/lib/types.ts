export type Fetch = typeof fetch;

export enum ButtonType {
	Up = -1,
	Down = 1
}

export type CreateProjectResponse = {
	id: number;
};

export type Curl = {
	id: number;
	name: string;
	description: string;
	rawQuery: string;
};

export type CurlGroup = {
	id: number;
	curls: string;
	description: string;
	labels: string;
	name: string;
	project_id: number;
};

export type Project = {
	info: ProjectInfo;
	admins: Array<string>;
	collaborators: Array<string>;
	groups: Array<CurlGroup>;
};

export type ProjectInfo = {
	id: number;
	environments: string;
	description: string;
	name: string;
	visibility: string;
};

export enum Visibility {
	Public = 'Public',
	Private = 'Private'
}

export const VisibilityOptions = [
	{
		name: 'visibility',
		value: 'Public',
		label: 'Public'
	},
	{
		name: 'visibility',
		value: 'Private',
		label: 'Private'
	}
];
