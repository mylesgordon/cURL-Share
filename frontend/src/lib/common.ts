export function splitAndTrim(input: string): Array<string> {
		return input.split(',').map((item) => item.trim());
	}