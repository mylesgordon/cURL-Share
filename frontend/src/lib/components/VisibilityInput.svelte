<script lang="ts">
	import { ChoiceInput } from 'agnostic-svelte';
	import { Visibility, VisibilityOptions } from '$lib/types';

	export let visibility: Visibility;

	let visibilityCheckedOption: Array<string> = [visibility as string];
	$: visibility = Visibility[visibilityCheckedOption[0] as Visibility];
</script>

<ChoiceInput
	id="visibility"
	type="radio"
	options={VisibilityOptions}
	legendLabel="Visibility"
	checkedOptions={visibilityCheckedOption}
	bind:checked={visibilityCheckedOption}
/>
