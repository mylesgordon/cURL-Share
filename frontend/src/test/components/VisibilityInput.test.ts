import { Visibility } from '$lib/types';
import { fireEvent, render, screen } from '@testing-library/svelte';
import { testMatchingSnapshot } from '../common';
import VisibilityInput from '$lib/components/VisibilityInput.svelte';
import type { ComponentProps } from 'svelte';

describe('VisibilityInput component', () => {
	const props: ComponentProps<VisibilityInput> = {
		visibility: Visibility.Public
	};
	testMatchingSnapshot(VisibilityInput, props);

	it.each([
		{ visibility: Visibility.Public, publicChecked: true, privateChecked: false },
		{ visibility: Visibility.Private, publicChecked: false, privateChecked: true }
	])(
		'$visibility should default with public button checked $publicChecked, private button checked $privateChecked',
		({ visibility, publicChecked, privateChecked }) => {
			render(VisibilityInput, { visibility });
			const publicButton: HTMLInputElement = screen.getByLabelText('Public');
			const privateButton: HTMLInputElement = screen.getByLabelText('Private');
			expect(publicButton.checked).toEqual(publicChecked);
			expect(privateButton.checked).toEqual(privateChecked);
		}
	);

	it('should be able to switch between public and private', () => {
		render(VisibilityInput, props);
		const publicButton: HTMLInputElement = screen.getByLabelText('Public');
		const privateButton: HTMLInputElement = screen.getByLabelText('Private');

		fireEvent.click(privateButton);
		expect(publicButton.checked).toEqual(false);
		expect(privateButton.checked).toEqual(true);

		fireEvent.click(publicButton);
		expect(publicButton.checked).toEqual(true);
		expect(privateButton.checked).toEqual(false);
	});
});
