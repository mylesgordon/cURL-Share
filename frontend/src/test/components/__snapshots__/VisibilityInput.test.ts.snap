// Vitest Snapshot v1

exports[`VisibilityInput component > should match snapshot 1`] = `
<body>
  <div>
    <fieldset
      class="radio-group s-K6jzjY3vjtlg"
    >
      <legend
        class="radio-legend s-K6jzjY3vjtlg"
      >
        Visibility
      </legend>
       
      <label
        class="radio-label-wrap s-K6jzjY3vjtlg"
      >
        <input
          class="radio s-K6jzjY3vjtlg"
          id="visibility-visibility-0"
          name="visibility"
          type="radio"
          value="Public"
        />
         
        <span
          aria-hidden="true"
          class="radio-label s-K6jzjY3vjtlg"
        />
         
        <span
          class="radio-label-copy s-K6jzjY3vjtlg"
        >
          Public
        </span>
         
      </label>
      <label
        class="radio-label-wrap s-K6jzjY3vjtlg"
      >
        <input
          class="radio s-K6jzjY3vjtlg"
          id="visibility-visibility-1"
          name="visibility"
          type="radio"
          value="Private"
        />
         
        <span
          aria-hidden="true"
          class="radio-label s-K6jzjY3vjtlg"
        />
         
        <span
          class="radio-label-copy s-K6jzjY3vjtlg"
        >
          Private
        </span>
         
      </label>
    </fieldset>
    <!--&lt;ChoiceInput&gt;-->
    <!--&lt;VisibilityInput&gt;-->
  </div>
</body>
`;
